### ISSUE TITLE
drift and scaling issue in VO

### ISSUE THREAD

**Issue by mvenka24** [Huangying-Zhan/DF-VO#5]
Hi,

Thank you for wonderful approach. 

while testing on our custom dataset, we were able see better result compared to DSO, ORB. However, we still observe drifting and scaling issue globally. we are thinking to  retrain monodepth2 and lightflownet by down sampling image sequence as possible approach to to resolve the issue.

Please find sample output attached. 
1 . they are the same road travailing in opposite direction. drift happens when have  180 degree camera rotation.
2,3: scale is file but there is 180 degree shift in VO. camera is traveling in opposite direction.

any suggestion or tips for better result on custom dataset?

Thanks,
Madhu 
![WhatsApp Image 2019-12-25 at 4 54 56 PM](https://user-images.githubusercontent.com/55761977/71451745-270cf080-2739-11ea-8a48-5e93834d1a09.jpeg)

**Comment by Huangying-Zhan** [comment]
hi @mvenka24 , Thank you for your interest in our work.

Regarding the scaling issue, our scale consistency mainly rely on a scale-consistent depth model and I believe it is better to retrain monodepth2 on your custom dataset since the depth model is still dataset/sceneario dependent. However, you may need to add the depth-consistency term we mentioned in the paper for monocular training. For stereo training, you don't need the depth-consistency term.

Regarding the rotation drifting issue. Our system will fail if there is no overlap between two frames (180 degree rotation). For small rotation, we have a simple mechanism to switch the tracking mode (Essential matrix / PnP). However, there is still a hard threshold to determine the mode which can sometimes fail for small motion/pure rotation. I am working on improving the system by having a better mechanism.

In general, I would suggest you to retrain the depth model first and see how it works. 
It is an easier job for getting improvement.
For the flow-network, I don't think you have to retrain it at the moment. We find that the liteflownet trained from a synthetic dataset is quite robust for most regions and our forward-backward consistency term can remove the incorrect optical flow predictions quite well.

Hope it helps.

**Comment by mvenka24** [comment]
Hi,

Thanks for your comments. training depth model with custom data didn't help much. 


while experimenting with different depth model listed in your repo, we observed that there is consistent drift in trajectory across different models while traveling back on the same path.

![map_custom_04_mono+stereo_640x192](https://user-images.githubusercontent.com/55761977/72386679-0a802a80-36df-11ea-8f52-784f91a04728.png)



for given image, trajectory for the same road should be parallel. I have updated intrinsic camera parameters,   tried on different resolution. but this behavior is consistent.   

our aim is to get accurate global pose of camera for complete trip. apart from retraining depth model, what else we can do to achieve this?


Thanks,
Madhu.
